docs/decisions/0008-migrating-to-new-learner-credit.md

```diff
@@ -0,0 +1,159 @@
+
+Through the first release of EMET, we are converting enterprise offers into EMET learner credit. We plan to migrate eligible enterprise offers customers over to use the EMET learner credit instead. However, this means the course page still needs to support coupon codes and subscription licenses. As such, the majority of the existing API calls and business logic already in place within the course page must remain until codes are phased out and subscription subsidies are supported by the EMET system.
+
+#### `CourseHeader` component
+
+The `CourseHeader` component is responsible for the display of the course title, image, related skills, and renders an "Enroll" CTA for each available course run. A course run is deemed "available" if course-discovery denotes the course run is `is_marketable: true`, `is_enrollable: true`, and is not archived.
+
+The existing `CourseRunCards` component is responsible for iterating through the available course runs and rendering a `CourseRunCard` component with the appropriate messaging and "Enroll" CTA (or "View course" CTA is learner is already enrolled) for each course run.
+
+The existing `CourseRunCard` (rendered by `CourseRunCards`) is responsible for figuring out the display text of the "Enroll" / "View course" CTA and renders the `EnrollAction` component. `EnrollAction` is what determines the *functionality* of the CTA depending on the type of subsidy being used to enroll (i.e., link to Data Sharing Consent, link to ecommerce basket page, disabled "Enroll" button, etc.).
+
+Having the "Enroll" CTA logic essentially split between 2 components (i.e., `CourseRunCard` and `EnrollAction`), it's increasingly difficult to reason about. 
+
+To mitigate this concern, we will deprecate the existing `CourseRunCards` component in favor of creating a more streamlined, lightweight `CourseRunCards` component instead. That is, we will have `CourseRunCards` that is integrated with the EMET APIs and fallback to `CourseRunCards.Deprecated` to remain backwards compatible.
+
+#### Triggering EMET redemption for a course run
+When a learner clicks on the "Enroll" CTA for a course run that is redeemable by the EMET system (i.e., has learner credit enabled with balance remaining), the general logic is as follows:
+* Make a `POST` request to the `redeem` endpoint for the redeemable access policy returned by `can_redeem`. This returns the transaction payload from `enterprise-subsidy`.
+* Poll against `enterprise-subsidy` API for the status of the returned transaction UUID until the transaction is in a non-pending state (e.g., "committed").
+  * If in a "committed" state, redirect learner to courseware URL for OCM courses to begin learning.
+  * If in an "error" state, we will ensure appropriate messaging is displayed and an option to retry.
```

Comment:
Currently the create transaction endpoint (POST /api/v1/transactions/) is synchronous.

https://github.com/openedx/enterprise-subsidy/blob/54bee5fb78367f2df72cd0fc809ae621dec88b36/enterprise_subsidy/apps/api/v1/views/transaction.py#L406-L407

Returns 201 (and state will be "committed"), or some other state (in which case the transaction may not get created, or it ended up in "failed").  Should we be changing this enterprise-subsidy endpoint to be async, or change this ADR to assume that it is synchronous?

Reply:
I guess in the GEAG case (when that is implemented), one possible response is 201 with "pending" transaction.  What makes the API call synchronous or asynchronous is really just whether the created transaction is "committed"/"failed" or "pending", which doesn't really change the fundamental shape of the endpoint.  Request args and status codes would not change when switching from synchronous to asynchronous mode.